### ISSUE TITLE
Add QIIME 2 plugin

### ISSUE THREAD

**Issue by gibsramen** [gibsramen/BIRDMAn#26]
With the upcoming release of [QIIME 2 2021.4](https://github.com/qiime2/qiime2/pull/576) the underlying version of Python is being updated from 3.6 to 3.8. Hopefully this should allow BIRDMAn to be converted into a valid plugin.

Need to consider semantic type & transformer for multi-dimensional arrays ([relevant Q2 forum post](https://forum.qiime2.org/t/thoughts-on-differential-and-tensor-types/18210) cc @mortonjt).

See [preview](https://forum.qiime2.org/t/preview-qiime-2-2021-4-development-changelog/18657/4) for additional details.

**Comment by mortonjt** [comment]
This is going to be very exciting! CC @ebolyen, @thermokarst

I've been prototyping out what the underlying types could look like in [q2-differential](https://github.com/mortonjt/q2-differential/tree/main/q2_differential).  Right now, I'm converging to xarray / arviz, since the netCDF format seems to fit the bill (see [_transformer.py](https://github.com/mortonjt/q2-differential/blob/main/q2_differential/_transformer.py)).  Futhermore, I've had some success with these types in a couple of different packages namely [q2-batch](https://github.com/mortonjt/q2-batch) and [q2-fido](https://github.com/mortonjt/q2-fido).

Regarding turning this into a pure qiime2 plugin, I'm not exactly sure what the best route is.  The problem is that MCMC is extremely compute hungry and process standard microbiome datasets without cluster support.  I personally found that dask-jobqueue can be extremely useful for parallelizing these processes (see PR [here](https://github.com/mortonjt/q2-differential/pull/15)); but the problem is that qiime2 doesn't currently play nicely with dask-jobqueue due to network issues that I don't currently understand. 

So I see 3 possible options to consider 
1. Create a qiime2 plugin that operates on a single taxon at a time; that way power hungry users can wrap qiime2 with slurm et al and parallelize as fast as the scheduler allows them to
2. Forget the qiime2 plugin_setup.py and just input / save qiime2 Artifacts directly and use cluster schedulers in the back end
3. Push to figure out how to integrate dask into qiime2

**Comment by mortonjt** [comment]
More thoughts on the qiime2 plugin, I think the more pressing issue is to finalize the types.

It'll be difficult to anticipate all of the possible use-cases of the FeatureTensor, but perhaps there are a few things we can set in stone, namely this tensor type
1. Can have optional `sampleid` axis
2. Is required to have `featureid` axis
3. Is required to have `monte_carlo_samples` axis

Of course, there can be multiple subtypes like `FeatureTensor[Differential]` that is required to have a `covariates` axis, or `FeatureTensor[Longitudinal]` is required to have a `time` axis and a `covariate` axis. From what I can tell xarray and/or arviz seems to fit the bill for this.  The key thing here is figuring out how to standardize the allowed vs optional axes.

**Comment by gibsramen** [comment]
I agree with (1) & (2) but I'm not sure about (3). Do we know that all use cases of `FeatureTensor` will involve MC sampling? I'm not super familiar with q2-micom but from the Q2 forum post @cdiener mentioned that fluxes are stored as `(sample) x (taxon) x (flux for each reaction)` which would seem applicable here.

Related to this I think `xarray.Dataset` would be better than `arviz.InferenceData` for generalizability.

**Comment by mortonjt** [comment]
@gibsramen , what do you think about refining this to `MonteCarloTensor` to make it explicitly for Bayesian computation? 
I'm suggesting this since we already have quite a few use cases with upcoming qiime2 Bayesian plugins; which we have a good handle on fleshing out the requirements.

I'm less familiar with q2-micom, but @cdiener feel free to comment and we can brainstorm additional tensor types.

**Comment by gibsramen** [comment]
Yeah I think `MonteCarloTensor` is a great idea to standardize specifically for the Bayesian stuff going on.
